qutip/solver/mcsolve.py

```diff
@@ -0,0 +1,311 @@
+            t_step, state = self._integrator.mcstep(t, copy=False)
+            norm = self.prob_func(state)
+            if norm <= self.target_norm:
+                self._do_collapse(norm_old, norm, t_old)
+                t_old, y_old = self._integrator.get_state(copy=False)
+                norm_old = 1.
+            else:
+                t_old, y_old = t_step, state
+                norm_old = norm
+
+        return t_old, _data.mul(y_old, 1 / self.norm_func(y_old))
+
+    def _do_collapse(self, norm_old, norm, t_prev):
+        """Find and apply a collapse."""
+        t_final, _ = self._integrator.get_state()
+        tries = 0
+        while tries < self.options.mcsolve['norm_steps']:
+            tries += 1
+            if (t_final - t_prev) < self.options.mcsolve['norm_t_tol']:
+                t_guess = t_final
+                _, state = self._integrator.get_state()
+                break
+            t_guess = (
+                t_prev
+                + ((t_final - t_prev)
+                   * np.log(norm_old / self.target_norm)
+                   / np.log(norm_old / norm))
+            )
+            if (t_guess - t_prev) < self.options.mcsolve['norm_t_tol']:
+                t_guess = t_prev + self.options.mcsolve['norm_t_tol']
+            _, state = self._integrator.mcstep(t_guess, copy=False)
+            norm2_guess = self.prob_func(state)
+            if (
+                np.abs(self.target_norm - norm2_guess) <
+                self.options.mcsolve['norm_tol'] * self.target_norm
+            ):
+                break
+            elif (norm2_guess < self.target_norm):
+                # t_guess is still > t_jump
+                t_final = t_guess
+                norm = norm2_guess
+            else:
+                # t_guess < t_jump
+                t_prev = t_guess
+                norm_old = norm2_guess
+
+        if tries >= self.options.mcsolve['norm_steps']:
+            raise Exception("Norm tolerance not reached. " +
+                            "Increase accuracy of ODE solver or " +
+                            "SolverOptions.mcsolve['norm_steps'].")
+
+        # t_guess, state is at the collapse
+        probs = np.zeros(len(self._n_ops))
+        for i, n_op in enumerate(self._n_ops):
+            probs[i] = n_op.expect_data(t_guess, state).real
+        probs = np.cumsum(probs)
+        which = np.searchsorted(probs, probs[-1] * self.generator.random())
+
+        state_new = self._c_ops[which].matmul_data(t_guess, state)
+        new_norm = self.norm_func(state_new)
+        if new_norm < self.options.mcsolve['mc_corr_eps']:
+            # This happen when the collapse is caused by numerical error
+            state_new = _data.mul(state, 1 / self.norm_func(state))
+        else:
+            state_new = _data.mul(state_new, 1 / new_norm)
+            self.collapses.append((t_guess, which))
+            self.target_norm = self.generator.random()
+        self._integrator.set_state(t_guess, state_new)
```

Comment:
I thought setting the state of an integrator in this way could not be done for some integrators (adams or bdf)? Does it not have an drawback (such as having to reset those integrators)?